### PR TITLE
Add support for 'MapPublicIpOnLaunch' subnet attribute in cloud/amazon/ec2_vpc_subnet

### PR DESCRIPTION
**Pull request by brandond**
##### SUMMARY
Adds support for MapPublicIpOnLaunch option on VPC subnets. Boto3 is required, as this option is not exposed by the legacy boto API.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
cloud/amazon/ec2_vpc_subnet

##### ANSIBLE VERSION
```
ansible 2.4.0
  config file = /home/ec2-user/repos/cmb-infosec-ansible/ansible.cfg
  configured module search path = [u'/home/ec2-user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ec2-user/ansible-env/lib/python2.7/site-packages/ansible
  executable location = /home/ec2-user/ansible-env/bin/ansible
  python version = 2.7.13+ (default, Feb 24 2017, 13:51:49) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

### CODE REVIEW

**Comment by brandond** [comment]
@willthames @s-hertel any chance I could get a review?

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py [hunk 1]
```diff
@@ -82,152 +89,140 @@
...
-from ansible.module_utils.basic import AnsibleModule
-from ansible.module_utils.ec2 import AnsibleAWSError, connect_to_aws, ec2_argument_spec, get_aws_connection_info
-
+    HAS_BOTO3 = False
 
-class AnsibleVPCSubnetException(Exception):
-    pass
 
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ec2 import *  # noqa
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: please make the imports explicit

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py [hunk 2]
```diff
@@ -82,152 +89,140 @@
...
+            conn.delete_tags(Resources=[subnet['id']], Tags=ansible_dict_to_boto3_tag_list(to_delete), DryRun=check_mode)
 
         to_add = dict((k, tags[k]) for k in tags if k not in cur_tags or cur_tags[k] != tags[k])
         if to_add:
-            vpc_conn.create_tags(resource_id, to_add, dry_run=check_mode)
+            conn.create_tags(Resources=[subnet['id']], Tags=ansible_dict_to_boto3_tag_list(to_add), DryRun=check_mode)
+
+    except ClientError as e:
+        if e.response['Error']['Code'] != "DryRunOperation":
+            module.fail_json(msg='Unable to update tags for {0}: {1}'.format(subnet['id'], e))
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: Use the [boto3 exception guidelines](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#boto3-2)

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py [hunk 3]
```diff
@@ -82,152 +89,140 @@
...
+        subnet = camel_dict_to_snake_dict(subnet)
 
-class AnsibleVPCSubnetDeletionException(AnsibleVPCSubnetException):
-    pass
+    if 'tags' in subnet:
+        subnet['tags'] = boto3_tag_list_to_ansible_dict(subnet['tags'])
+    else:
+        subnet['tags'] = dict()
 
+    if 'subnet_id' in subnet:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:121]: Nice, I was a bit worried the output might change but I see you've thought about that with this.

**Review by willthames — CHANGES_REQUESTED** [review]
Just minor things really, shouldn't be too huge a blocker

**Comment by willthames** [comment]
I've run this on a test EC2 VPC and it seems to work exactly as I'd expect, so no concerns over functionality.

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
Closing and reopening for hopefully one more last CI run

**Pull request merged**